File: docs/absorbed-from-phenotype-gfx/unity/terrain/src/ChunkMeshBuilder.cs
using System;
using UnityEngine;

namespace Phenotype.Terrain
{
    /// <summary>
    /// Holds the output of a mesh build operation: vertices, triangle indices,
    /// UV coordinates and per-vertex normals.
    /// </summary>
    /// <remarks>
    /// All arrays are parallel: the element at index <c>i</c> in <see cref="Vertices"/>
    /// corresponds to the same vertex in <see cref="UVs"/> and <see cref="Normals"/>.
    /// </remarks>
    /// <example>
    /// <code>
    /// var builder = new ChunkMeshBuilder();
    /// MeshData mesh = builder.BuildMesh(16, 1.0f);
    ///
    /// // Upload to Unity
    /// var unityMesh = new Mesh();
    /// unityMesh.vertices = mesh.Vertices;
    /// unityMesh.triangles = mesh.Indices;
    /// unityMesh.uv = mesh.UVs;
    /// unityMesh.normals = mesh.Normals;
    /// </code>
    /// </example>
    public class MeshData
    {
        /// <summary>
        /// Vertex positions in world space.
        /// </summary>
        /// <remarks>The length equals the total number of vertices in the mesh.</remarks>
        public Vector3[] Vertices { get; set; }

        /// <summary>
        /// Triangle index buffer (3 indices per triangle).
        /// </summary>
        /// <remarks>
        /// Each consecutive group of 3 integers defines one triangle in clockwise winding order
        /// when viewed from above (positive Y).
        /// </remarks>
        public int[] Indices { get; set; }

        /// <summary>
        /// UV coordinates for each vertex.
        /// </summary>
        /// <remarks>U and V range from 0 to 1 across the chunk.</remarks>
        public Vector2[] UVs { get; set; }

        /// <summary>
        /// Per-vertex normals.
        /// </summary>
        /// <remarks>For flat meshes this defaults to <c>Vector3.up</c>.</remarks>
        public Vector3[] Normals { get; set; }
    }

    /// <summary>
    /// Generates Unity Mesh objects for terrain chunks from height-field data.
    /// Handles vertex layout, triangle winding, UV mapping, and normals.
    /// </summary>
    /// <remarks>
    /// <para>Two overloads are provided:</para>
    /// <list type="bullet">
    /// <item><description><see cref="BuildMesh(int, float)"/> produces a flat grid at Y = 0.</description></item>
    /// <item><description><see cref="BuildMesh(HeightField, int, float)"/> samples elevation from a <see cref="HeightField"/>.</description></item>
    /// </list>
    /// </remarks>
    /// <example>
    /// <code>
    /// var builder = new ChunkMeshBuilder();
    ///
    /// // Flat 16x16 grid, 100 units wide
    /// MeshData flat = builder.BuildMesh(16, 100f);
    ///
    /// // Height-mapped grid
    /// var heightField = new HeightField(17, 17);
    /// MeshData terrain = builder.BuildMesh(heightField, 16, 100f);
    /// </code>
    /// </example>
    public class ChunkMeshBuilder
    {
        /// <summary>
        /// Builds a flat mesh for a terrain chunk with the given grid resolution.
        /// The chunk spans [0, size] in both X and Z, with Y = 0.
        /// </summary>
        /// <param name="resolution">Number of grid quads along each axis. Must be &gt; 0.</param>
        /// <param name="size">World-space size of the chunk. Default is 1.</param>
        /// <returns>A <see cref="MeshData"/> containing the generated mesh.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when resolution is &lt;= 0.</exception>
        /// <example>
        /// <code>
        /// var builder = new ChunkMeshBuilder();
        /// MeshData mesh = builder.BuildMesh(32, 200f);
        ///
        /// // mesh.Vertices.Length == (32 + 1) * (32 + 1) == 1089
        /// // mesh.Indices.Length == 32 * 32 * 2 * 3 == 6144
        /// </code>
        /// </example>
        public MeshData BuildMesh(int resolution, float size = 1f)
        {
            if (resolution <= 0)
                throw new ArgumentOutOfRangeException(nameof(resolution), "resolution must be > 0");

            int vertexCount = (resolution + 1) * (resolution + 1);
            int triangleCount = resolution * resolution * 2;
            int indexCount = triangleCount * 3;

            var vertices = new Vector3[vertexCount];
            var indices = new int[indexCount];
            var uvs = new Vector2[vertexCount];
            var normals = new Vector3[vertexCount];

            float cellSize = size / resolution;

            // Generate vertices and UVs
            for (int z = 0; z <= resolution; z++)
            {
                for (int x = 0; x <= resolution; x++)
                {
                    int index = z * (resolution + 1) + x;
                    vertices[index] = new Vector3(x * cellSize, 0f, z * cellSize);
                    uvs[index] = new Vector2((float)x / resolution, (float)z / resolution);
                    normals[index] = new Vector3(0f, 1f, 0f);
                }
            }

            // Generate indices with consistent clockwise winding
            int idx = 0;
            for (int z = 0; z < resolution; z++)
            {
                for (int x = 0; x < resolution; x++)
                {
                    int bottomLeft = z * (resolution + 1) + x;
                    int bottomRight = bottomLeft + 1;
                    int topLeft = (z + 1) * (resolution + 1) + x;
                    int topRight = topLeft + 1;

                    // First triangle (clockwise when viewed from above)
                    indices[idx++] = bottomLeft;
                    indices[idx++] = topLeft;
                    indices[idx++] = topRight;

                    // Second triangle (clockwise when viewed from above)
                    indices[idx++] = bottomLeft;
                    indices[idx++] = topRight;
                    indices[idx++] = bottomRight;
                }
            }

            return new MeshData
            {
                Vertices = vertices,
                Indices = indices,
                UVs = uvs,
                Normals = normals,
            };
        }

        /// <summary>
        /// Builds a mesh for a terrain chunk using the supplied height field.
        /// Y values are sampled from <paramref name="heightField"/>.
        /// </summary>
        /// <param name="heightField">Height field to sample elevation from.</param>
        /// <param name="resolution">Number of grid quads along each axis. Must be &gt; 0.</param>
        /// <param name="size">World-space size of the chunk. Default is 1.</param>
        /// <returns>A <see cref="MeshData"/> containing the generated mesh.</returns>
        /// <exception cref="ArgumentOutOfRangeException">Thrown when resolution is &lt;= 0.</exception>
        /// <example>
        /// <code>
        /// var heightField = new HeightField(33, 33);
        /// for (int x = 0; x &lt; 33; x++)
        ///     for (int z = 0; z &lt; 33; z++)
        ///         heightField.SetHeight(x, z, Mathf.Sin(x * 0.1f) * 10f);
        ///
        /// var builder = new ChunkMeshBuilder();
        /// MeshData mesh = builder.BuildMesh(heightField, 32, 100f);
        /// </code>
        /// </example>
        public MeshData BuildMesh(HeightField heightField, int resolution, float size = 1f)
        {
            if (resolution <= 0)
                throw new ArgumentOutOfRangeException(nameof(resolution), "resolution must be > 0");

            // For now, HeightField is a stub with no data; fall back to flat mesh.
            return BuildMesh(resolution, size);
        }
    }
}

File: docs/absorbed-from-phenotype-gfx/src/terrain/chunk_mesh_builder.rs
//! Flat + height-mapped chunk mesh builder.
//!
//! Ported from C# `ChunkMeshBuilder.cs`. The `BuildMesh(HeightField, int, float)`
//! overload was a stub that delegated to `BuildMesh(int, float)` and is dropped
//! per the audit (YAGNI).
//!
//! The output `MeshData` uses `glam::Vec3` for positions/normals and
//! `glam::Vec2` for UVs.

use serde::{Deserialize, Serialize};

use crate::terrain::error::{TerrainError, TerrainResult};
use crate::terrain::height_field::HeightField;

/// Vertex + index + UV + normal buffer for a chunk mesh.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MeshData {
    /// Vertex positions in world space.
    pub vertices: Vec<glam::Vec3>,
    /// Triangle index buffer (3 indices per triangle, clockwise when viewed from above).
    pub indices: Vec<u32>,
    /// Per-vertex UV coordinates.
    pub uvs: Vec<glam::Vec2>,
    /// Per-vertex normals (default: `Vec3::Y` for flat meshes).
    pub normals: Vec<glam::Vec3>,
}

/// Generates a chunk mesh from a height field (or flat grid).
#[derive(Debug, Default, Clone, Copy)]
pub struct ChunkMeshBuilder;

impl ChunkMeshBuilder {
    /// Build a flat grid mesh at Y = 0.
    ///
    /// `resolution` is the number of quads per side; `size` is the world-space
    /// edge length. The output has `(resolution + 1)²` vertices and
    /// `resolution² * 6` indices.
    pub fn build_mesh(&self, resolution: i32, size: f32) -> TerrainResult<MeshData> {
        if resolution <= 0 {
            return Err(TerrainError::InvalidResolution { value: resolution });
        }
        let res_u = resolution as usize;
        let vertex_count = (res_u + 1) * (res_u + 1);
        let index_count = res_u * res_u * 6;

        let mut vertices = Vec::with_capacity(vertex_count);
        let mut uvs = Vec::with_capacity(vertex_count);
        let mut normals = Vec::with_capacity(vertex_count);

        let cell_size = size / resolution as f32;
        for z in 0..=res_u {
            for x in 0..=res_u {
                vertices.push(glam::Vec3::new(x as f32 * cell_size, 0.0, z as f32 * cell_size));
                uvs.push(glam::Vec2::new(x as f32 / resolution as f32, z as f32 / resolution as f32));
                normals.push(glam::Vec3::Y);
            }
        }

        let mut indices = Vec::with_capacity(index_count);
        for z in 0..res_u {
            for x in 0..res_u {
                let bottom_left = (z * (res_u + 1) + x) as u32;
                let bottom_right = bottom_left + 1;
                let top_left = ((z + 1) * (res_u + 1) + x) as u32;
                let top_right = top_left + 1;
                // First triangle (clockwise from above)
                indices.push(bottom_left);
                indices.push(top_left);
                indices.push(top_right);
                // Second triangle (clockwise from above)
                indices.push(bottom_left);
                indices.push(top_right);
                indices.push(bottom_right);
            }
        }

        Ok(MeshData { vertices, indices, uvs, normals })
    }

    /// Build a height-mapped mesh by sampling `height_field` at each grid vertex.
    /// `height_field` must have dimensions `(resolution + 1) x (resolution + 1)`.
    pub fn build_mesh_from_height(&self, height_field: &HeightField, resolution: i32, size: f32) -> TerrainResult<MeshData> {
        if resolution <= 0 {
            return Err(TerrainError::InvalidResolution { value: resolution });
        }
        let res_u = resolution as usize;
        let vertex_count = (res_u + 1) * (res_u + 1);
        let index_count = res_u * res_u * 6;
        let expected_w = res_u as i32 + 1;
        let expected_h = res_u as i32 + 1;
        if height_field.width() != expected_w || height_field.height() != expected_h {
            return Err(TerrainError::InvalidDataLength {
                got: (height_field.width() * height_field.height()) as usize,
                expected: expected_w as usize * expected_h as usize,
            });
        }

        let mut vertices = Vec::with_capacity(vertex_count);
        let mut uvs = Vec::with_capacity(vertex_count);
        let mut normals = Vec::with_capacity(vertex_count);

        let cell_size = size / resolution as f32;
        for z in 0..=res_u {
            for x in 0..=res_u {
                let y = height_field.get_height(x as i32, z as i32).unwrap_or(0.0);
                vertices.push(glam::Vec3::new(x as f32 * cell_size, y, z as f32 * cell_size));
                uvs.push(glam::Vec2::new(x as f32 / resolution as f32, z as f32 / resolution as f32));
                normals.push(glam::Vec3::Y);
            }
        }

        let mut indices = Vec::with_capacity(index_count);
        for z in 0..res_u {
            for x in 0..res_u {
                let bottom_left = (z * (res_u + 1) + x) as u32;
                let bottom_right = bottom_left + 1;
                let top_left = ((z + 1) * (res_u + 1) + x) as u32;
                let top_right = top_left + 1;
                indices.push(bottom_left);
                indices.push(top_left);
                indices.push(top_right);
                indices.push(bottom_left);
                indices.push(top_right);
                indices.push(bottom_right);
            }
        }

        Ok(MeshData { vertices, indices, uvs, normals })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn build_mesh_simple_quad_produces_correct_counts() {
        let mesh = ChunkMeshBuilder.build_mesh(1, 1.0).unwrap();
        assert_eq!(mesh.vertices.len(), 4);
        assert_eq!(mesh.uvs.len(), 4);
        assert_eq!(mesh.normals.len(), 4);
        assert_eq!(mesh.indices.len(), 6);
    }

    #[test]
    fn build_mesh_simple_quad_vertices_at_expected_positions() {
        let mesh = ChunkMeshBuilder.build_mesh(1, 2.0).unwrap();
        assert_eq!(mesh.vertices[0], glam::Vec3::new(0.0, 0.0, 0.0));
        assert_eq!(mesh.vertices[1], glam::Vec3::new(2.0, 0.0, 0.0));
        assert_eq!(mesh.vertices[2], glam::Vec3::new(0.0, 0.0, 2.0));
        assert_eq!(mesh.vertices[3], glam::Vec3::new(2.0, 0.0, 2.0));
    }

    #[test]
    fn build_mesh_simple_quad_indices_form_two_triangles() {
        let mesh = ChunkMeshBuilder.build_mesh(1, 1.0).unwrap();
        // First triangle: bottom-left(0), top-left(2), top-right(3)
        assert_eq!(mesh.indices[0], 0);
        assert_eq!(mesh.indices[1], 2);
        assert_eq!(mesh.indices[2], 3);
        // Second triangle: bottom-left(0), top-right(3), bottom-right(1)
        assert_eq!(mesh.indices[3], 0);
        assert_eq!(mesh.indices[4], 3);
        assert_eq!(mesh.indices[5], 1);
    }

    #[test]
    fn build_mesh_triangle_winding_order_is_upward() {
        let mesh = ChunkMeshBuilder.build_mesh(4, 4.0).unwrap();
        let triangle_count = mesh.indices.len() / 3;
        for t in 0..triangle_count {
            let i0 = mesh.indices[t * 3] as usize;
            let i1 = mesh.indices[t * 3 + 1] as usize;
            let i2 = mesh.indices[t * 3 + 2] as usize;
            let v0 = mesh.vertices[i0];
            let v1 = mesh.vertices[i1];
            let v2 = mesh.vertices[i2];
            let edge1 = v1 - v0;
            let edge2 = v2 - v0;
            let normal = edge1.cross(edge2);
            assert!(normal.y > 0.0, "triangle {t} has downward-facing normal");
        }
    }

    #[test]
    fn build_mesh_vertex_count_near_resolution() {
        let mesh = ChunkMeshBuilder.build_mesh(64, 64.0).unwrap();
        // (64+1)² = 4225
        assert_eq!(mesh.vertices.len(), 4225);
    }

    #[test]
    fn build_mesh_vertex_count_mid_resolution() {
        let mesh = ChunkMeshBuilder.build_mesh(32, 64.0).unwrap();
        // (32+1)² = 1089
        assert_eq!(mesh.vertices.len(), 1089);
    }

    #[test]
    fn build_mesh_invalid_resolution() {
        assert!(ChunkMeshBuilder.build_mesh(0, 1.0).is_err());
        assert!(ChunkMeshBuilder.build_mesh(-1, 1.0).is_err());
    }
}
